Sources/_SwiftSyntaxTestSupport/FixItApplier.swift

```diff
@@ -0,0 +1,123 @@
+    filterByMessages messages: [String]?,
+    to tree: any SyntaxProtocol
+  ) -> String {
+    let messages = messages ?? diagnostics.compactMap { $0.fixIts.first?.message.message }
+
+    let changes =
+      diagnostics
+      .flatMap(\.fixIts)
+      .filter { messages.contains($0.message.message) }
+      .flatMap(\.changes)
+
+    var edits: [Edit] = changes.edits
+    var source = tree.description
+
+    while let edit = edits.first {
+      edits = Array(edits.dropFirst())
+
+      let startIndex = source.utf8.index(source.utf8.startIndex, offsetBy: edit.startUtf8Offset)
+      let endIndex = source.utf8.index(source.utf8.startIndex, offsetBy: edit.endUtf8Offset)
+
+      source.replaceSubrange(startIndex..<endIndex, with: edit.replacement)
+
+      edits = edits.compactMap { remainingEdit -> FixItApplier.Edit? in
+        var remainingEdit = remainingEdit
+
+        if remainingEdit.replacementRange.overlaps(edit.replacementRange) {
+          // The edit overlaps with the previous edit. We can't apply both
+          // without conflicts. Apply the one that's listed first and drop the
+          // later edit.
+          return nil
+        }
+
+        // If the remaining edit starts after or at the end of the edit that we just applied,
+        // shift it by the current edit's difference in length.
+        if edit.endUtf8Offset <= remainingEdit.startUtf8Offset {
+          // Take the largest value as we in some cases can end up with a negative offset.
+          remainingEdit.startUtf8Offset = remainingEdit.startUtf8Offset - edit.replacementRange.count + edit.replacementLength
+
+          // If we need to replace a string at the end, we need to ensure we don't get an out-of-bounds error.
+          remainingEdit.endUtf8Offset = remainingEdit.endUtf8Offset - edit.replacementRange.count + edit.replacementLength
+        }
+
+        return remainingEdit
+      }
+    }
+
+    return source
+  }
+}
+
+fileprivate extension FixIt.Change {
+  var edit: FixItApplier.Edit {
+    switch self {
+    case .replace(let oldNode, let newNode):
+      return FixItApplier.Edit(
+        startUtf8Offset: oldNode.position.utf8Offset,
+        endUtf8Offset: oldNode.endPosition.utf8Offset,
+        replacement: newNode.description
+      )
+
+    case .replaceLeadingTrivia(let token, let newTrivia):
+      return FixItApplier.Edit(
+        startUtf8Offset: token.position.utf8Offset,
+        endUtf8Offset: token.positionAfterSkippingLeadingTrivia.utf8Offset,
+        replacement: newTrivia.description
+      )
+
+    case .replaceTrailingTrivia(let token, let newTrivia):
+      return FixItApplier.Edit(
+        startUtf8Offset: token.endPositionBeforeTrailingTrivia.utf8Offset,
+        endUtf8Offset: token.endPosition.utf8Offset,
+        replacement: newTrivia.description
+      )
+    }
+  }
+}
+
+fileprivate extension Array where Element == FixIt.Change {
+  var edits: [FixItApplier.Edit] {
+    return self.map { $0.edit }
+  }
+}
```

Comment:
Do we really need this? Wouldn’t it be simpler if we just do `changes.map { $0.edit }` inline above?